fdbserver/DDTeamCollection.actor.cpp

```diff
@@ -3234,36 +3273,89 @@ class DDTeamCollectionImpl {
 	}
 }; // class DDTeamCollectionImpl
 
-void DDTeamCollection::updatePivotAvailableSpaceRatio() {
-	if (now() - lastPivotAvailableSpaceUpdate > SERVER_KNOBS->AVAILABLE_SPACE_UPDATE_DELAY) {
-		lastPivotAvailableSpaceUpdate = now();
-		std::vector<double> teamAvailableSpace;
-		teamAvailableSpace.reserve(teams.size());
-		for (const auto& team : teams) {
-			if (team->isHealthy()) {
-				teamAvailableSpace.push_back(team->getMinAvailableSpaceRatio());
-			}
+void DDTeamCollection::updateAvailableSpacePivots() {
+	std::vector<double> teamAvailableSpace;
+	for (int i = 0; i < teams.size(); ++i) {
+		if (teams[i]->isHealthy()) {
+			teamAvailableSpace.push_back(teams[i]->getMinAvailableSpaceRatio());
 		}
+	}
 
-		size_t pivot = teamAvailableSpace.size() * std::min(1.0, SERVER_KNOBS->AVAILABLE_SPACE_PIVOT_PERCENT);
-		if (teamAvailableSpace.size() > 1) {
-			std::nth_element(teamAvailableSpace.begin(), teamAvailableSpace.begin() + pivot, teamAvailableSpace.end());
-			pivotAvailableSpaceRatio =
-			    std::max(SERVER_KNOBS->MIN_AVAILABLE_SPACE_RATIO,
-			             std::min(SERVER_KNOBS->TARGET_AVAILABLE_SPACE_RATIO, teamAvailableSpace[pivot]));
-		} else {
-			pivotAvailableSpaceRatio = SERVER_KNOBS->MIN_AVAILABLE_SPACE_RATIO;
+	if (!teamAvailableSpace.empty()) {
+		size_t pivot = teamAvailableSpace.size() * std::min(1.0, SERVER_KNOBS->AVAILABLE_SPACE_PIVOT_RATIO);
+		std::nth_element(
+		    teamAvailableSpace.begin(), teamAvailableSpace.begin() + pivot, teamAvailableSpace.end(), std::greater{});
+		teamPivots.pivotAvailableSpaceRatio =
+		    std::max(SERVER_KNOBS->MIN_AVAILABLE_SPACE_RATIO,
+		             std::min(SERVER_KNOBS->TARGET_AVAILABLE_SPACE_RATIO, teamAvailableSpace[pivot]));
+	} else {
```

Comment:
There is a potential issue which may exist for a long time here. 

Saying we set `SERVER_KNOBS->AVAILABLE_SPACE_PIVOT_RATIO >= 1.0`. then `size_t pivot = teamAvailableSpace.size()`, we would then access `teamAvailableSpace[pivot] = teamAvailableSpace[ teamAvailableSpace.size()]`, which is out of the boundary of vector.

CPU Pivot below also has the same issue. Maybe we need to add a safeguard in `ServerKnobs.cpp` like `ASSERT_LT(AVAILABLE_SPACE_PIVOT_RATIO, 1.0)`

Reply:
Added in #9964